python/services/dataservice/dmod/dataservice/service.py

```diff
@@ -1,25 +1,321 @@
+
+        Parameters
+        ----------
+        message : DatasetManagementMessage
+            The incoming message, expected to include data to be added to a dataset.
+        mngr : DatasetManager
+            The manager instance for the relevant dataset.
+
+        Returns
+        -------
+        DatasetManagementResponse
+            Generated response to the manager message for adding data.
+
+        See Also
+        -------
+        ::method:`_async_process_add_data`
+        """
+        if not isinstance(message, DatasetManagementMessage):
+            return DatasetManagementResponse(success=False, reason="Unparseable Message Received")
+        elif message.management_action != ManagementAction.ADD_DATA:
+            msg_txt = "Expected {} action but received {}".format(ManagementAction.ADD_DATA, message.management_action)
+            return DatasetManagementResponse(success=False, reason="Unexpected Management Action", message=msg_txt)
+        elif message.data is None:
+            return DatasetManagementResponse(success=False, reason="No Data In ADD_DATA Message")
+        elif mngr.add_data(message.dataset_name, dest=message.data_location, data=message.data):
+            return DatasetManagementResponse(success=True, reason="Data Added", is_awaiting=message.is_pending_data)
+        else:
+            return DatasetManagementResponse(success=False, reason="Failure Adding Data To Dataset", is_awaiting=False)
+
+    def _process_dataset_create(self, message: DatasetManagementMessage) -> DatasetManagementResponse:
+        """
+        As part of the communication protocol for the service, handle incoming messages that request dataset creation.
+
+        Parameters
+        ----------
+        message : DatasetManagementMessage
+            The message that initiated the process of creating a new dataset
+
+        Returns
+        ----------
+        DatasetManagementResponse
+            A generated response object to the incoming creation message, indicating whether creation was successful.
+        """
+        # Make sure there is no conflict/existing dataset already
+        if message.dataset_name in self.get_known_datasets():
+            return DatasetManagementResponse(success=False, reason="Dataset Already Exists")
+        # Handle when message to create fails to include a dataset domain
+        elif message.data_domain is None:
+            msg = "Invalid {} for dataset creation: no dataset domain provided.".format(self.__class__.__name__)
+            return DatasetManagementResponse(success=False, reason="No Dataset Domain", message=msg)
+
+        # Create the dataset
+        dataset_type = self._determine_dataset_type(message)
+        self._all_data_managers[dataset_type].create(name=message.dataset_name, category=message.data_category,
+                                                     domain=message.data_domain, is_read_only=False)
```

Comment:
I think this is part of something that will need to be revisited in the future, but basically because, why create the dataset with nothing in it, then make it read-only?